src/js/components/invoice/AddItemsPage.jsx

```diff
@@ -337,6 +362,46 @@ class AddItemsPage extends Component {
       });
   }
 
+  validate() {
+    const errors = {};
+    errors.invoiceItems = [];
+
+    _.forEach(this.state.values.invoiceItems, (item, key) => {
+      if (_.has(item, 'isValid') && !item.isValid) {
+        errors.invoiceItems[key] = { quantity: 'react.invoice.errors.quantityToInvoice.label' };
+      }
+    });
+
+    return errors;
+  }
+
+  async validateInvoiceItem({
+    invoiceItemId,
+    quantity,
+    rowIndex,
+  }) {
+    this.debouncedInvoiceItemValidation.cancel();
+
+    const { data } = await invoiceApi.validateInvoiceItem({
+      invoiceItemId,
+      quantity,
+    });
+
+    const mappedInvoiceItems = this.state.values.invoiceItems.map((item, index) => {
+      if (index === rowIndex) {
+        return { ...item, quantity, isValid: data.isValid };
+      }
+      return item;
+    });
+
+    this.setState(previousState => ({
+      values: {
+        ...previousState.values,
+        invoiceItems: mappedInvoiceItems,
+      },
+    }));
```

Comment:
use the `immutability-helper` it handles such nested state changes and will shorten the code and make it much more readable.
```js
const newValues = update(this.state.values, {
  invoiceItems: {
     [rowIndex]: { isValid: { $set:  data.isValid } }
  }
})
this.setState({ values: newValues })
```